metaci/api/views/testmethod_perf.py

```diff
@@ -0,0 +1,284 @@
+                if filtername in self.disable_by_default:
+                    self.filters[filtername] = copy.copy(self.filters[filtername])
+                    self.filters[filtername].method = self.dummy_filter
+
+    def dummy_filter(self, queryset, name, value):
+        return queryset
+
+
+class BuildFlowFilterSet(TurnFilterSetOffByDefaultBase):
+    """
+       The tricky bit is that this filter serves three different jobs.
+
+        1. It validates incoming parameters (which is not strictly necessary
+          but probably difficult to turn off and perhaps sometimes useful).
+
+        2. It populates the django-filter form.
+
+        3. It drives actual filtering of the build-list sub-query.
+
+        Feature 3 is turned on and off by the really_filter parameter.
+        We do NOT want django-filter to try to automatically filter the
+        output queryset based on these filters because it is actually the
+        sub-query that we need to filter.
+
+        Accordingly, its fields are turned off by default (see disable_by_default) 
+        and turned on explicitly ("really_filter") when it is created by get_queryset
+       """
+
+    build_fields = {
+        "repo": "build__repo__name",
+        "plan": "build__plan__name",
+        "branch": "build__branch__name",
+        "flow": "flow",
+    }
+
+    repo_choices = (
+        Repository.objects.values_list("name", "name").order_by("name").distinct()
+    )
+    repo = django_filters.rest_framework.ChoiceFilter(
+        field_name="build__repo__name", label="Repo Name", choices=repo_choices
+    )
+
+    branch_choices = (
+        Branch.objects.values_list("name", "name").order_by("name").distinct()
+    )
+    branch = django_filters.rest_framework.ChoiceFilter(
+        field_name="build__branch__name", label="Branch Name", choices=branch_choices
+    )
+
+    plan_choices = Plan.objects.values_list("name", "name").order_by("name").distinct()
+    plan = django_filters.rest_framework.ChoiceFilter(
+        field_name="build__plan__name", label="Plan Name", choices=plan_choices
+    )
+
+    flow_choices = (
+        BuildFlow.objects.values_list("flow", "flow").order_by("flow").distinct()
+    )
+    flow = django_filters.rest_framework.ChoiceFilter(
+        field_name="flow", label="Flow Name", choices=flow_choices
+    )
+
+    recentdate = DateRangeFilter(label="Recent Date", field_name="time_end")
+
+    daterange = django_filters.rest_framework.DateFromToRangeFilter(
+        field_name="time_end",
+        label="Date range",
+        widget=DateRangeWidget(attrs={"type": "date"}),
+    )
+
+    # This is not really a filter. It's actually just a query input but putting it
+    # here lets me get it in the form.
+    build_flows_limit = django_filters.rest_framework.NumberFilter(
+        method="dummy_filter", label="Build Flows Limit (default: 100)"
+    )
+
+    disable_by_default = dir()  # disable all of these fields by default.
```

Comment:
I didn't realize that `dir()` could be used without an argument to list the names of locals, so thanks for the comment. I'm not entirely convinced it's better than being explicit.

Reply:
I'll try a decorator and see if that looks better. I do kind of want to preserve the property that adding a method to this class will "automatically" give it the conditional behaviour. But a decorator might be clearer about where and how the magic is happening.

Reply:
@davisagli What do you think about it as a decorator?

https://github.com/SFDO-Tooling/MetaCI/blob/06ee51b2b9625c4cbab2d86b343e7fc5aa6746b7/metaci/api/views/testmethod_perf.py#L59